Minor improvements to feature/scan-compile-run

d3adb5 opened the pull request:
Some minor incremental improvements to the scan-compile-run flow, seeking to
improve legibility and, despite the "big design upfront" sin, allow the code to
be more smoothly extended in the future.

Proposed changes:

  - Return a function from `CommandFacade::parse()` instead of executing it.
  - Display a usage message if `./main` is invoked with no arguments.
  - Add a `list_commands()` method to list available commands and use it.

Currently a draft, pending some discussion and a little more work.

d3adb5 committed:
refactor(facade): return function instead of running it

Return the function that executes a given command instead of simply
executing it if it is found. This is done for the sake of clarity and
single responsibility: should a function called 'parse' yield the result
of said parsing or DO something as a result of said parsing?

IMHO, this improves legibility.

d3adb5 committed:
feat: display usage message when given no args

Display a basic usage message when the command is given no arguments
upon being invoked. Currently very basic, but sets the stage for
extension later on.

d3adb5 committed:
feat: list available commands in usage message

List the available commands in the usage message by turning the
CommandFacade instance into a global variable, initialized before
main(), and accessing the list_commands() function from it.

To ensure a possible destructor method is called upon exit, an atexit
handler (delete_facade) was introduced that invokes the delete
operation.

d3adb5 commented:
I'll keep discussion about these changes and future ones I'm planning in this PR, if that's okay.

Is there any reason why `CommandFacade` needs to be instantiable (non-abstract), @vaugusto92? Or at the very least, for `assert`, `commands`, `parse()`, etc. to be instance members instead of class members? If they can be `static`, we can avoid having to instantiate `CommandFacade` where we need it.

I know an `Assert` instance is used in `CommandFacade`, but given it's a singleton, what are the practical benefits of instantiating here? ELI5, I'm not an OOP advocate or experienced user.

d3adb5 commented:
Here's an idea: in [5a83dc4](https://github.com/vaugusto92/cpp-assert/pull/7/commits/5a83dc4f4cbd18a556c26abb4795f4e4a3888762) I made it so `CommandFacade::parse()` has a return type of `std::function<void()>`; I'm thinking of changing that to `std::function<int()>`, so the commands that are executed can have control of the exit status for the program.

It would help indicate failure (and potentially what type of failure) when running specific commands. I'm not sure if there's a use case for that just yet, so I guess this too is BDUF.

What do you think?

d3adb5 commented:
This might be a radical thought, and considering you're a Clojure developer we might both be biased in this direction, but what do you think of making an effort to eliminate (well, refactor) as many of the methods / functions that return `void`, @vaugusto92?

vaugusto92 commented:
Hello, Mr. Wilson!

Concerning this idea:

> Is there any reason why `CommandFacade` needs to be instantiable (non-abstract), @vaugusto92? Or at the very least, for `assert`, `commands`, `parse()`, etc. to be instance members instead of class members? If they can be `static`, we can avoid having to instantiate `CommandFacade` where we need it.

You're absolutely right. It should be either a Singleton or an "abstract class", or even better - a namespace with the functions we need.

vaugusto92 commented:
About this idea:

> Here's an idea: in [5a83dc4](https://github.com/vaugusto92/cpp-assert/pull/7/commits/5a83dc4f4cbd18a556c26abb4795f4e4a3888762) I made it so `CommandFacade::parse()` has a return type of `std::function<void()>`; I'm thinking of changing that to `std::function<int()>`, so the commands that are executed can have control of the exit status for the program.
> 
> It would help indicate failure (and potentially what type of failure) when running specific commands. I'm not sure if there's a use case for that just yet, so I guess this too is BDUF.
> 
> What do you think?

That was needed. In fact, it would be better for the `parse` function to have control of what happened in the test executions. Idea approved :)

vaugusto92 commented:
To this:

> This might be a radical thought, and considering you're a Clojure developer we might both be biased in this direction, but what do you think of making an effort to eliminate (well, refactor) as many of the methods / functions that return `void`, @vaugusto92?

It is not radical, but it requires some caution. I think we might eliminate as many as we can, but for some of them it might be complicated to find a meaningful return.

d3adb5 committed:
refactor: convert command_facade into namespace

Make command_facade a namespace instead of an instantiable or abstract
class. It isn't a data type, after all, so there's no sense in treating
it as such.

d3adb5 commented:
I had a go at converting `Assert` into a namespace instead of keeping it as a class. Call it OOP prejudice or whatever, but it didn't sit right with me to create a data type for something that is inherently an utility aggregate. Sure, this logic can be applied to perhaps 90% of classes you see in big projects written in OOP languages, but... they're inherently OOP. I digress.

Anyway, I managed to concoct some inane, malformed, cursed C++ code in the end that seemed to work, albeit throwing multiple warnings when running `./main test`. I did not dare bundle it together with this PR, so if you want to give it a look, please do: [e17c431](https://github.com/d3adb5/cpp-assert/commit/e17c431cafaf6c6a6e473ddf8501cc2403d6f1db)

I am marking this PR as ready for review because I imagine you haven't shown the `feature/scan-compile-run` a lot of love lately on my account. Let's get the PR done with so you can return to writing benign C++ while I repeat to myself countless times just how grateful I am to not be a C++ developer.

vaugusto92 commented:
> I had a go at converting `Assert` into a namespace instead of keeping it as a class. Call it OOP prejudice or whatever, but it didn't sit right with me to create a data type for something that is inherently an utility aggregate. Sure, this logic can be applied to perhaps 90% of classes you see in big projects written in OOP languages, but... they're inherently OOP. I digress.
> 
> Anyway, I managed to concoct some inane, malformed, cursed C++ code in the end that seemed to work, albeit throwing multiple warnings when running `./main test`. I did not dare bundle it together with this PR, so if you want to give it a look, please do: [e17c431](https://github.com/d3adb5/cpp-assert/commit/e17c431cafaf6c6a6e473ddf8501cc2403d6f1db)
> 
> I am marking this PR as ready for review because I imagine you haven't shown the `feature/scan-compile-run` branch a lot of love lately on my account. Let's get the PR done with so you can return to writing benign C++ while I repeat to myself countless times just how grateful I am to not be a C++ developer.

Thank you for the contribution! Question. Did you run `./main test` inside the container or in your machine?

d3adb5 commented:
> Thank you for the contribution! Question. Did you run `./main test` inside the container or in your machine?

I ran it on my machine, hoping there would be little to no difference when using the G++ version running on Alpine. I just compiled and ran the tests inside the container, however, and all seems well.

That said, shouldn't the GitHub Actions workflow have kicked in when I marked the PR as ready for review? Or even when creating it as a draft...

d3adb5 committed:
chore: trigger ci workflow on pull requests

Add the generic pull request event trigger to the CI workflow, so every
PR comes with it as a status check.

d3adb5 commented:
I took the liberty of adding a `pull_request` trigger to the CI workflow alongside this PR. Hope that's okay with you. I found it odd from the beginning that there weren't any status checks when I opened this. :open_mouth:

vaugusto92 commented:
> I took the liberty of adding a `pull_request` trigger to the CI workflow alongside this PR. Hope that's okay with you. I found it odd from the beginning that there weren't any status checks when I opened this. open_mouth

Thank you, sir! It was set to "push" only. I asked about your environment because I ran the `./main test` command inside the container and it didn't prompt me any warnings at all haha I think it is fine like that. :)

d3adb5 commented:
> Thank you, sir! It was set to "push" only. I asked about your environment because I ran the `./main test` command inside the container and it didn't prompt me any warnings at all haha I think it is fine like that. :)

Try this:

```bash
git remote add fork https://github.com/d3adb5/cpp-assert
git fetch fork && git checkout e17c431
docker build . -t cppzinho
docker run -it --rm -v $(pwd):/work cppzinho
```

And _then_ `make; ./main test`.
